src/www/ui/api/Controllers/LicenseController.php

```diff
@@ -84,14 +104,227 @@ public function getLicense($request, $response, $args)
+      }
+      if ($page > $totalPages) {
+        $info = new Info(400, "Can not exceed total pages: $totalPages",
+          InfoType::ERROR);
+        $retVal = $response->withHeader("X-Total-Pages", $totalPages)
+          ->withJson($info->getArray(), $info->getCode());
+      }
+    } else {
+      $page = 1;
+    }
+    if ($retVal !== null) {
+      return $retVal;
+    }
+    $onlyActive = $request->getHeaderLine(self::ACTIVE_PARAM);
+    if (! empty($onlyActive)) {
+      $onlyActive = filter_var($onlyActive, FILTER_VALIDATE_BOOLEAN);
+    } else {
+      $onlyActive = false;
+    }
+
+    $licenses = $this->dbHelper->getLicensesPaginated($page, $limit,
+      $this->restHelper->getGroupId(), $onlyActive);
+    $licenseList = [];
+
+    foreach ($licenses as $license) {
+      $newRow = new License(
+        $license['rf_pk'],
+        $license['rf_shortname'],
+        $license['rf_fullname'],
+        $license['rf_text'],
+        $license['rf_url'],
+        null,
+        $license['rf_risk'],
+        $license['group_fk'] != 0
+      );
+      $licenseList[] = $newRow->getArray();
+    }
+
+    return $response->withHeader("X-Total-Pages", $totalPages)
+      ->withJson($licenseList, 200);
+  }
+
+  /**
+   * Create a new license
+   *
+   * @param Request $request
+   * @param Response $response
+   * @param array $args
+   * @return Response
+   */
+  public function createLicense($request, $response, $args)
+  {
+    $newLicense = $request->getParsedBody();
+    $newLicense = License::parseFromArray($newLicense);
+    $newInfo = null;
+    if ($newLicense === -1) {
+      $newInfo = new Info(400, "Input contains additional properties.",
+        InfoType::ERROR);
+    } elseif ($newLicense === -2) {
+      $newInfo = new Info(400, "Property 'shortName' is required.",
+        InfoType::ERROR);
+    } elseif (! $newLicense->getIsCandidate() && ! Auth::isAdmin()) {
+      $newInfo = new Info(403, "Need to be admin to create non-candidate " .
+        "license.", InfoType::ERROR);
+    }
+    if ($newInfo !== null) {
+      return $response->withJson($newInfo->getArray(), $newInfo->getCode());
+    }
+    $tableName = "license_ref";
+    $assocData = [
+      "rf_shortname" => $newLicense->getShortName(),
+      "rf_fullname" => $newLicense->getFullName(),
+      "rf_text" => $newLicense->getText(),
+      "rf_risk" => $newLicense->getRisk(),
+      "rf_url" => $newLicense->getUrl(),
+      "rf_detector_type" => 1
+    ];
+    if ($newLicense->getIsCandidate()) {
+      $tableName = "license_candidate";
+      $assocData["group_fk"] = $this->restHelper->getGroupId();
+      $assocData["rf_user_fk_created"] = $this->restHelper->getUserId();
+      $assocData["rf_user_fk_modified"] = $this->restHelper->getUserId();
+      $assocData["marydone"] = $newLicense->getMergeRequest();
+    }
+    $rfPk = $this->dbHelper->getDbManager()->insertTableRow($tableName,
+      $assocData, __METHOD__ . ".newLicense", "rf_pk");
```

Comment:
duplicate candidate license check require before adding a new candidate license.

Reply:
@ag4ums I have added checks for duplicate shortname for both main license and candidate license. Also added a check for duplicate license text for main license.